From 071afad56fcc1f10633ebd7c93767d3005b35e50 Mon Sep 17 00:00:00 2001
From: Ronan Abhamon <ronan.abhamon@vates.fr>
Date: Tue, 24 Oct 2023 23:07:23 +0200
Subject: [PATCH 141/171] fix(LinstorSR): wait during attach to open DRBD path

ENODATA and other errors like EROFS can be raised when
a new DRBD path is created on the fly. So ensure to block before tapdisk starts.

Signed-off-by: Ronan Abhamon <ronan.abhamon@vates.fr>
---
 drivers/LinstorSR.py | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/LinstorSR.py b/drivers/LinstorSR.py
index d0ce079..3063abf 100755
--- a/drivers/LinstorSR.py
+++ b/drivers/LinstorSR.py
@@ -1901,6 +1901,25 @@ class LinstorVDI(VDI.VDI):
                 raise xs_errors.XenError(
                     'VDIUnavailable', opterr='Could not find: {}'.format(path)
                 )
+
+            # Diskless path can be created on the fly, ensure we can open it.
+            def check_volume_usable():
+                while True:
+                    try:
+                        with open(path, 'r+'):
+                            pass
+                    except IOError as e:
+                        if e.errno == errno.ENODATA:
+                            time.sleep(2)
+                            continue
+                        if e.errno == errno.EROFS:
+                            util.SMlog('Volume not attachable because RO. Openers: {}'.format(
+                                self.sr._linstor.get_volume_openers(vdi_uuid)
+                            ))
+                        raise
+                    break
+            util.retry(check_volume_usable, 15, 2)
+
             vdi_uuid = self.sr._vhdutil.get_vhd_info(vdi_uuid).parentUuid
 
         self.attached = True
-- 
2.44.0

